From e35bdfc692b43f3dd3ca0f265529026d84217aff Mon Sep 17 00:00:00 2001
From: David Barratt <dbarratt@wikimedia.org>
Date: Fri, 18 May 2018 16:25:37 +0200
Subject: [PATCH] Allow the vendor directory to be configurable.

The location of the vendor directory can be configured in Composer, but cannot
be configured inside of MediaWiki. The vendor directory should be configurable
so users can store the vendor directory outside of the webroot. To configure
the the vendor directory, load the autoloader in the front controller and
MediaWiki will skip loading it.

Bug: T194880
Change-Id: I416ae0521fbb52f6150727ceeb3325a364e3cdfe
---

diff --git a/composer.json b/composer.json
index 6e34ec2..94c1e76 100644
--- a/composer.json
+++ b/composer.json
@@ -108,7 +108,8 @@
 	},
 	"config": {
 		"optimize-autoloader": true,
-		"prepend-autoloader": false
+		"prepend-autoloader": false,
+		"autoloader-suffix": "_mediawiki_vendor"
 	},
 	"extra": {
 		"merge-plugin": {
diff --git a/includes/PHPVersionCheck.php b/includes/PHPVersionCheck.php
index cfe889f..5e2610c 100644
--- a/includes/PHPVersionCheck.php
+++ b/includes/PHPVersionCheck.php
@@ -151,7 +151,10 @@
 	 * Displays an error, if the vendor/autoload.php file could not be found.
 	 */
 	function checkVendorExistence() {
-		if ( !file_exists( dirname( __FILE__ ) . '/../vendor/autoload.php' ) ) {
+		if (
+			!class_exists( 'ComposerAutoloaderInit_mediawiki_vendor', false )
+			&& !file_exists( dirname( __FILE__ ) . '/../vendor/autoload.php' )
+		) {
 			$shortText = "Installing some external dependencies (e.g. via composer) is required.";
 
 			$longText = "Error: You are missing some external dependencies. \n"
diff --git a/includes/Setup.php b/includes/Setup.php
index 5cc9a96..00ffa8c 100644
--- a/includes/Setup.php
+++ b/includes/Setup.php
@@ -54,8 +54,11 @@
 // Load global functions
 require_once "$IP/includes/GlobalFunctions.php";
 
-// Load composer's autoloader if present
-if ( is_readable( "$IP/vendor/autoload.php" ) ) {
+// Load composer's autoloader if present and not already loaded.
+if (
+	!class_exists( 'ComposerAutoloaderInit_mediawiki_vendor', false )
+	&& is_readable( "$IP/vendor/autoload.php" )
+) {
 	require_once "$IP/vendor/autoload.php";
 }
 
